XenCenterLib/Archive/ArchiveWriter.cs

```diff
@@ -52,29 +52,51 @@ public virtual void SetBaseStream(Stream outputStream)
 
         public void CreateArchive(string pathToArchive, Action cancellingDelegate = null, Action<int> progressDelegate = null)
         {
-            if (!Directory.Exists(pathToArchive))
+            if (!Directory.Exists(StringUtility.ToLongWindowsPath(pathToArchive)))
                 throw new FileNotFoundException("The path " + pathToArchive + " does not exist");
 
-            var files = Directory.GetFiles(pathToArchive, "*.*", SearchOption.AllDirectories);
-            for (var i = 0; i < files.Length; i++)
+            // We look recursively and do not use Directory.GetDirectories and Directory.GetFiles with
+            // AllDirectories options because in .NET 4.8 they do not enumerate all elements if they
+            // have paths longer than 260 characters (240 for directories).
+            // If moving to .NET Core, please consider reverting this .
+            AddFiles(StringUtility.ToLongWindowsPath(pathToArchive), cancellingDelegate);
+            var directories = Directory.GetDirectories(StringUtility.ToLongWindowsPath(pathToArchive));
+            for (var j = 0; j < directories.Length; j++)
+            {
+                string dirPath = directories[j];
+                cancellingDelegate?.Invoke();
+                AddDirectory(CleanRelativePathName(pathToArchive, dirPath), Directory.GetCreationTime(dirPath));

```

Comment:
I think these methods can be compacted a bit to one recursive method. Could we perhaps rename the existing abstarct AddDirectory to AddDirectoryCore and then create a new private one which will be calling AddDirectoryCore (upon condition, as I understand that we don't want to include the top directory), then the content of AddFiles, then the recursive AddDirectory on each subfolder. The cancelling delegate should be easy to call in there, although calculating the progress may be slightly trickier.